add fork to child_process

sirenkovladd opened the pull request:
Copy from https://github.com/nodejs/node/blob/v20.5.0/lib/child_process.js#L101-L173
and corrected for `bun`

Jarred-Sumner commented:
did you try it and see if it works?

sirenkovladd commented:
Fixes https://github.com/oven-sh/bun/issues/3771 https://github.com/oven-sh/bun/issues/3620

sirenkovladd commented:
> did you try it and see if it works?

Yes, for `dev` of course)
currently generating for release, may take some time

Jarred-Sumner commented:
How does it communicate with the parent process and support sending data? The native code parts don't support additional file descriptors I don't think. Or is that feature less commonly used and this is more used as a way to run a file?

birkskyum commented on src/js/out/modules/node/child_process.js:
```diff
@@ -705,7 +745,6 @@ var child_process_default = {
 export {
   spawnSync,
   spawn,
-  fork,
```
How come the named export is removed?

sirenkovladd replied:
I lost the export on the function, i fixed it

Jarred-Sumner commented:
Can you include automated tests? These can be copied from Node. See https://github.com/oven-sh/bun/blob/main/test/js/node/child_process/child_process-node.test.js#L21  for an example

sirenkovladd commented:
Ok, actually I found the blocks for this PR, it's the `process.send` method

Without it I can't send messages between processes

But I think this PR can be accepted and we can make a issue on solve `process.send` / `process.on('message')`

sirenkovladd commented:
### What does this PR do?

Implement method `fork` in `child_process`

the fork method also needs methods for handling messages (process.send, process.on('message'), ChildProcess.send) to be fully compatible with node

- [X] Documentation or TypeScript types (it's okay to leave the rest blank in this case)
- [X] Code changes

### How did you verify your code works?

<!-- **For code changes, please include automated tests**. Feel free to uncomment the line below -->

- [X] I wrote automated tests
- [X] I tested in my project which used `fork` from `node`

If JavaScript/TypeScript modules or builtins changed:

- [ ] I ran `make js` and committed the transpiled changes
(No rule to make target 'js')
- [X] I or my editor ran Prettier on the changed files (or I ran `bun fmt`)
- [X] I included a test for the new code, or an existing test covers it

<!-- If Zig files changed:

- [ ] I checked the lifetime of memory allocated to verify it's (1) freed and (2) only freed when it should be
- [ ] I or my editor ran `zig fmt` on the changed files
- [ ] I included a test for the new code, or an existing test covers it
- [ ] JSValue used outside outside of the stack is either wrapped in a JSC.Strong or is JSValueProtect'ed
-->

If new methods, getters, or setters were added to a publicly exposed class:

- [ ] I added TypeScript types for the new methods, getters, or setters
- [X] The types already existed

If dependencies in tests changed:

- [X] I made sure that specific versions of dependencies are used instead of ranged or tagged versions


<!-- If functions were added to exports.zig or bindings.zig

- [ ] I ran `make headers` to regenerate the C header file

-->

<!-- If \*.classes.ts files were added or changed:

- [ ] I ran `make codegen` to regenerate the C++ and Zig code
-->

<!-- If a new builtin ESM/CJS module was added:

- [ ] I updated Aliases in `module_loader.zig` to include the new module
- [ ] I added a test that imports the module
- [ ] I added a test that require() the module
-->

Jarred-Sumner commented on test/js/node/child_process/child_process-node.test.js:
```diff
@@ -483,3 +492,261 @@ describe("child_process double pipe", () => {
+  describe.todo("stdio", () => {
+    // https://github.com/nodejs/node/blob/v20.5.0/test/parallel/test-child-process-fork-stdio.js
+  });
+  describe.only("fork", () => {
```
It should not be using .only, instead .skip

sirenkovladd replied:
Fix it, thank you for finding this

Jarred-Sumner commented:
This looks good, minus the usage of describe.only

Jarred-Sumner commented:
The test for message should just be removed entirely since none of that works and a test which immediately calls done should be skipped since it does nothing and it won’t be entirely clear to future readers why the test exists or what it was supposed to do 

That being said, this is an improvement over the status quo so thank you and I’ll merge it now
